### DISCUSSION TITLE
Limit no of Agents

### DISCUSSION THREAD

**Discussion by ghost** [wazuh/wazuh#18556] [Q&A]
Hi together,

I'm looking for a solution to limit the no of agents per instance. I know there was previously a Env Var that allowed limiting the number of agents but this was deprecated.
Is there any way to limit the number of agents?

Thank you very much for your assistance

**Comment by pereyra-m** [comment]
Hello @iFrozenPhoenix !

I'd need more information to understand your use case.
It's true that there isn't a specific setting to limit the number of agents, but the agents can't register against the manager if the administrator doesn't allow this, so you are always in control of the enrollment process.

The agents are able to request a key by themselves, but this behavior can be disabled (see [enrollment](https://documentation.wazuh.com/current/user-manual/reference/ossec-conf/client.html#enrollment)). 
And even if they try to register, the registration daemon `wazuh-authd` must be running to generate a new key. You can turn off this module in the configuration (see [auth](https://documentation.wazuh.com/current/user-manual/reference/ossec-conf/auth.html)).

So please, give me more details about your environment.
Regards.

**Reply by ghost** [reply]
Hello @pereyra-m .

Thank you for your response. We are using Wazuh in our corporate environment. All our Clients and Servers are registered in our Inventory management system. We would like to limit the number of Wazuh Agents to the number of Clients and Servers in our Inventory management system, to be more exact, to certain groups of Servers and Clients.
We want to do this to better allocate the internal Wazuh costs to the organizational units. Currently we allocate the internal costs by checking the Wazuh API for the registered Clients. But we would like to limit the possible Clients in advance by setting a limit of connected Agents.

Regards

**Reply by pereyra-m** [reply]
Thank you for the details @iFrozenPhoenix !

What do you think of using the [Agent key request](https://documentation.wazuh.com/current/user-manual/agents/key-request.html#agent-key-request) feature?

It requires a script that retrieves the keys from an external source (like a DB), and it could fail in case a specific limit is reached.

Please, give it a try and come back with any trouble you have during the implementation.

Regards.

**Comment by ghost** [comment]
@pereyra-m Thank you for the hint with the agent key request feature. After a bit of integration work it now works as intended.

Regards

**Reply by saad-zaib** [reply]
Hi @vexvec  i am working on something similar. Can you help me out on how you would have done this, as I am looking to integrate this feature to limit the number of agents being added to the manager

**Comment by pereyra-m** [comment]
Thank you for the update!
Regards.
